tests/fv/flannel_migration_test.go

```diff
@@ -0,0 +1,381 @@
+	backend "github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	client "github.com/projectcalico/libcalico-go/lib/clientv3"
+	"github.com/projectcalico/libcalico-go/lib/ipam"
+	cnet "github.com/projectcalico/libcalico-go/lib/net"
+	"github.com/projectcalico/libcalico-go/lib/options"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	// Flannel migration controller will run on this node.
+	controllerNodeName = "node-0"
+	// Master node Name.
+	masterNodeName = "node-1"
+
+	flannelDs = "kube-flannel-ds-amd64"
+
+	flannelSubnetEnv = "FLANNEL_NETWORK=192.168.0.0/16;FLANNEL_SUBNET=192.168.1.1/24;FLANNEL_MTU=8951;FLANNEL_IPMASQ=true;"
+)
+
+var (
+	emptyLabel = map[string]string{}
+)
+
+var _ = Describe("flannel-migration-controller FV test", func() {
+	var (
+		etcd                *containers.Container
+		migrationController *containers.Container
+		apiserver           *containers.Container
+		calicoClient        client.Interface
+		bc                  backend.Client
+		k8sClient           *kubernetes.Clientset
+		controllerManager   *containers.Container
+		kconfigfile         *os.File
+		err                 error
+		flannelCluster      *testutils.FlannelCluster
+	)
+
+	logKubectl := func(args ...string) {
+		out, err := apiserver.ExecOutput(args...)
+		Expect(err).ShouldNot(HaveOccurred())
+		logrus.Infof("--- kubectl output --- \n%s", out)
+	}
+
+	startController := func() {
+		// Add 3 seconds delay before main thread starts, this is to make sure FV can add watch channels
+		// before controller logs out to stderr.
+		// (container.Run returns after 'docker ps' shows the container, the polling interval is 1 second.)
+		// Add 60 seconds delay before main thread exits, this is to make sure controller is still running
+		// after test case completed and stopped by AfterEach.
+		migrationController = testutils.RunFlannelMigrationController(kconfigfile.Name(), controllerNodeName, flannelSubnetEnv, 3, 60)
+	}
+
+	stopController := func() {
+		migrationController.Stop()
+	}
+
+	BeforeEach(func() {
+		// Run etcd.
+		etcd = testutils.RunEtcd()
+
+		// Run apiserver.
+		apiserver = testutils.RunK8sApiserver(etcd.IP)
+
+		// Write out a kubeconfig file
+		kconfigfile, err = ioutil.TempFile("", "ginkgo-migrationcontroller")
+		Expect(err).NotTo(HaveOccurred())
+
+		data := fmt.Sprintf(testutils.KubeconfigTemplate, apiserver.IP)
+		_, err = kconfigfile.Write([]byte(data))
+		Expect(err).NotTo(HaveOccurred())
+
+		k8sClient, err = testutils.GetK8sClient(kconfigfile.Name())
+		Expect(err).NotTo(HaveOccurred())
+
+		// Wait for the apiserver to be available.
+		Eventually(func() error {
+			_, err := k8sClient.CoreV1().Namespaces().List(metav1.ListOptions{})
+			return err
+		}, 30*time.Second, 1*time.Second).Should(BeNil())
+
+		// Apply the necessary CRDs.
+		out, err := apiserver.ExecOutput("kubectl", "apply", "-f", "crds.yaml")
```

Comment:
We seem to support a CRDS_FILE environment variable.

https://github.com/projectcalico/kube-controllers/blob/acc821422ef66904c52298e8b3cfcfc78e59faad/tests/testutils/utils.go#L58

I don't seem to see where we're using crds.yaml for our other tests.... do they not need this file?

Reply:
I think we do use crds.yaml for every test cases. Migration controller FV creates api-server and apply `crds.yaml` in `BeforeEach` for all test cases.
https://github.com/song-jiang/kube-controllers/blob/song-f2c-fv/tests/fv/flannel_migration_test.go#L122